Enable inquiry/mode page data load from file

XiaowenJiang opened the pull request:
1. Add command line argument "page_file" in scsi hd property.

    2. Load page file at scsi harddisk initialization and save in SCSIDiskState.

    3. Add process in mode_sense/inquiry commands to read user-insert page data.

Signed-off-by: Julie Jiang <julie.jiang@dell.com>

weia5-emc committed:
Enable inquiry/mode page data load from file

    1. Add command line argument "page_file" in scsi hd property.

    2. Load page file at scsi harddisk initialization and save in SCSIDiskState.

    3. Add process in mode_sense/inquiry commands to read user-insert page data.

Signed-off-by: Julie Jiang <julie.jiang@dell.com>

xiar commented on hw/scsi/scsi-disk.c:
```diff
@@ -570,12 +573,68 @@ static uint8_t *scsi_get_buf(SCSIRequest *req)
+typedef struct __file
+{
+    uint16_t inquiry_offset;
+    uint16_t mode_offset;
```
i didn't see where the 'mode_offset' was referenced.  inquiry_offset and mode_offset are the same?  the file layout is as below, correct?
+------------------------------------------------------------------------------------------------------------------+
 |    file header |  pages header | page index | page data | page index | page data | ... | pages header | page index | ...  |
+------------------------------------------------------------------------------------------------------------------+

xiar replied:
page type == 0 is inquiry page page type ==  1 is the mode page?

emcinfrasim replied:
Yes, page type == 0 is inquiry and 1 is mode page.
The file layout is as below:
| file header | inquiry page header | inq page1 | inq page2| ...| standard inq page | mode page header | mode page1 | mode page2 | ...| checksum | 

In the page header, the structure is:
| number of pages |  page index1 | page index2 | ... |

emcinfrasim replied:
妈呀我用错账号了！

xiar commented on hw/scsi/scsi-disk.c:
```diff
@@ -570,12 +573,68 @@ static uint8_t *scsi_get_buf(SCSIRequest *req)
+            buflen = page_load(s, req->cmd.buf[2], 0, &outbuf, 0, &available_space);
+        } else {
+            // Standard Page
+            buflen = page_load(s, 0xff, 0xff, &outbuf, 0, &available_space);
```
in the emulation data, subpage and page are 0xff?

emcinfrasim replied:
No, because there is no page code for standard page, and we know that INQUIRY Pages have no subpages. This is the reason we use page code = 0xff, sub page code = 0xff for Standard Inquiry page, so that we could put it into "inquiry" group in our generated bin file without create another space.

xiar commented on hw/scsi/scsi-disk.c:
```diff
@@ -570,12 +573,68 @@ static uint8_t *scsi_get_buf(SCSIRequest *req)
+    uint8_t sub_page_code;
+    uint16_t offset;
+    uint16_t length;
+}PageIndex;
```
it's better to follow the qemu coding style.

emcinfrasim replied:
will do

XiaowenJiang committed:
Beautify code

1. add comments.
2. refact variable name.

Signed-off-by: Aaron Wei <aaron.wei@dell.com>

weia5-emc committed:
Modify page emulation commands to combine fetched data with command-line parameters

1. Modify inquiry page emulation to fill page 0x80, 0x83, 0xb0, 0xb2 with command-line arguments.
2. Add page code: 0x80, 0x83, 0xb0, 0xb2 in VPD inquiry page 0x00.
3. Modify mode page emulation to fill page 0x01, 0x04, 0x05, 0x08 with command-line arguments.

Signed-off-by: Julie Jiang <julie.jiang@dell.com>

xiar reviewed (approved):
👍
